plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/SpringReload.java

```diff
@@ -0,0 +1,572 @@
+                }
+            }
+        }
+    }
+
+    private void clearCache() {
+        // spring won't rebuild dependency map if injectionMetadataCache is not cleared
+        // which lead to singletons depend on beans in xml won't be destroy and recreate, may be a spring bug?
+        ResetSpringStaticCaches.reset();
+        ResetBeanPostProcessorCaches.reset(beanFactory);
+        ResetTransactionAttributeCaches.reset(beanFactory);
+        ResetBeanFactoryPostProcessorCaches.reset(beanFactory);
+        ProxyReplacer.clearAllProxies();
+        // fixme temperately disable it
+//        ResetBeanFactoryCaches.reset(beanFactory);
+        ConfigurationClassPostProcessorEnhance.getInstance(beanFactory).postProcess(beanFactory);
+        ResetAnnotationCache.resetAnnotationScanner(beanFactory);
+    }
+
+    private void reloadAnnotatedBeanDefinitions(Class clazz, String[] beanNames, List<String> needReloadBeanDefinitionNames) {
+        for (String beanName : beanNames) {
+            if (beanName.startsWith("&")) {
+                beanName = beanName.substring(1);
+            }
+            BeanDefinition beanDefinition = BeanFactoryProcessor.getBeanDefinition(beanFactory, beanName);
+            if (AnnotationHelper.hasAnnotation(clazz, "org.springframework.context.annotation.Configuration")
+                    && beanDefinition.getAttribute("org.springframework.context.annotation.ConfigurationClassPostProcessor.configurationClass") != null) {
+                needReloadBeanDefinitionNames.add(beanName);
+                String generateBeanName = beanNameGenerator.generateBeanName(beanDefinition, beanFactory);
+                // the beanName is not the same as generateBeanName, it should register a new bean definition and remove the old one.
+                if (!beanName.equals(generateBeanName)) {
+                    beanFactory.removeBeanDefinition(beanName);
+                    BeanDefinition newBeanDefinition = new AnnotatedGenericBeanDefinition(clazz);
+                    beanFactory.registerBeanDefinition(generateBeanName, newBeanDefinition);
+                } else {
+                    // remove the attribute, and it will recreate the bean
+                    ((AbstractBeanDefinition) beanDefinition).setBeanClass(clazz);
+                    beanDefinition.removeAttribute("org.springframework.context.annotation.ConfigurationClassPostProcessor.configurationClass");
+                }
+            }
+        }
+    }
+
+    private void destroyBeans() {
+        for (String beanName : new ArrayList<>(recreatedBeans)) {
+            destroyBean(beanName);
+        }
+        // destroy factory bean
+        for (String beanName : beanFactory.getBeanDefinitionNames()) {
+            BeanDefinition beanDefinition = beanFactory.getBeanDefinition(beanName);
+            if (beanDefinition.getFactoryMethodName() != null && beanDefinition.getBeanClassName() != null &&
```

Comment:
Is it possible to refactor the judgements in 'if' and in 'else if' into a seperated method and call it 'isCreatedByFactoryAndNeedReloead()'?

Reply:
Good IDEA